From 2241aa7f3a9dca8cea76f6e01c7e9e7fc6504439 Mon Sep 17 00:00:00 2001
From: Vinu Vaghasia <vinu.vaghasia@amd.com>
Date: Tue, 23 Nov 2021 15:14:41 -0600
Subject: [PATCH 1/1] u-boot-aspeed-sdk : Modify board_id and dts bug fix

- main.c : board_id code changed to convert the upper case letter
- image-fit.c : code added to convert board_id string to hex

Signed-off-by: Vinu Vaghasia <vinu.vaghasia@amd.com>
---
 common/image-fit.c | 40 +++++++++++++++++++++++-----------------
 common/main.c      | 23 ++++++++++++++---------
 2 files changed, 37 insertions(+), 26 deletions(-)

diff --git a/common/image-fit.c b/common/image-fit.c
index 61746e841c..e2000c5ea1 100644
--- a/common/image-fit.c
+++ b/common/image-fit.c
@@ -31,6 +31,7 @@ DECLARE_GLOBAL_DATA_PTR;
 #include <u-boot/sha1.h>
 #include <u-boot/sha256.h>
 #include <u-boot/sha512.h>
+#include <linux/ctype.h>

 /*****************************************************************************/
 /* New uImage format routines */
@@ -1786,6 +1787,23 @@ static const char *fit_get_image_type_property(int type)
 	return "unknown";
 }

+#define HEX_BASE	16
+#define DECIMAL_BASE	10
+
+uint32_t str2hex(char *str)
+{
+	uint32_t val = 0;
+
+	for (int i = 0; i < strlen(str); i++) {
+		if (isxdigit(str[i])) {
+			if (isdigit(str[i])) str[i] = str[i] - '0';
+			else str[i] = toupper(str[i]) - 'A'+ DECIMAL_BASE;
+		}
+		val = (val * HEX_BASE) | (str[i] & 0xF);
+	}
+	return val;
+}
+
 int fit_image_load(bootm_headers_t *images, ulong addr,
 		   const char **fit_unamep, const char **fit_uname_configp,
 		   int arch, int image_type, int bootstage_id,
@@ -1844,32 +1862,20 @@ int fit_image_load(bootm_headers_t *images, ulong addr,
 			// Reading board_id from ENV
 			platform_id = env_get("board_id");
 			// Convert bd_id string to integer
-			board_id = ( (platform_id[0]-'0')*16 + (platform_id[1]-'0'));
-
+			board_id = str2hex(platform_id);
 			// select the  device tree configuration based on board_id
 			switch (board_id) {
 				case 0x3D:
-				case 0x40:
-				case 0x41:
-				case 0x42:
+				case 0x40 ... 0x42:
 				case 0x52:
 					sprintf((char *)fit_uname_config, "conf-aspeed-bmc-amd-onyx.dtb"); break;
 				case 0x3E:
-				case 0x43:
-				case 0x44:
-				case 0x45:
+				case 0x43 ... 0x45:
 				case 0x51:
 					sprintf((char *)fit_uname_config, "conf-aspeed-bmc-amd-quartz.dtb"); break;
-				case 0x46:
-				case 0x47:
-				case 0x48:
+				case 0x46 ... 0x48:
 					sprintf((char *)fit_uname_config, "conf-aspeed-bmc-amd-ruby.dtb"); break;
-				case 0x49:
-				case 0x4A:
-				case 0x4B:
-				case 0x4C:
-				case 0x4D:
-				case 0x4E:
+				case 0x49 ... 0x4E:
 					sprintf((char *)fit_uname_config, "conf-aspeed-bmc-amd-titanite.dtb"); break;
 				default :
 					sprintf((char *)fit_uname_config, "conf-aspeed-bmc-amd-quartz.dtb"); break;
diff --git a/common/main.c b/common/main.c
index eb3cd5254c..b5e119b203 100644
--- a/common/main.c
+++ b/common/main.c
@@ -12,13 +12,14 @@
 #include <console.h>
 #include <version.h>
 #include <i2c.h>
+#include <linux/ctype.h>

-
-#define EEPROM_I2C_BUS 7
-#define EEPROM_DEV_ADDR 0X50
-#define EEPROM_BOARD_ID_OFFSET 0X10
-#define EEPROM_ADDR_LEN 2
-#define BOARD_ID_LEN 1
+#define EEPROM_I2C_BUS			7
+#define EEPROM_DEV_ADDR		0X50
+#define EEPROM_BOARD_ID_OFFSET		0X10
+#define EEPROM_ADDR_LEN		2
+#define BOARD_ID_LEN			1
+#define BOARD_ID_BUFF_LEN		2

 /*
  * Board-specific Platform code can reimplement show_boot_progress () if needed
@@ -32,8 +33,8 @@ void set_board_id()
 {
 	char *s;
 	unsigned char board_id [BOARD_ID_LEN];
-	int ret;
-	char brdid[2];
+	int ret, i=0;
+	char brdid[BOARD_ID_BUFF_LEN];

 	/* Check if the board_id env is set */
 	s = env_get("board_id");
@@ -45,9 +46,13 @@ void set_board_id()
 		printf("Error : Retrieving board_id from EEPROM Failed\n");
 	}
 	else {
-		/* if the env is not set OR env id  and eeprom id are different, then set from eeprom id */
 		sprintf(brdid, "%x", board_id[0]);
+		/* Convert char to upper case */
+		for ( i=0; i < BOARD_ID_BUFF_LEN; i++) {
+			brdid[i]= toupper(brdid[i]);
+		}

+		/* if the env is not set OR env id  and eeprom id are different, then set from eeprom id */
 		if ( (s == NULL) || (strcmp(s, brdid) !=0 ) ) {
 			env_set("board_id", brdid);
 			env_save();
--
2.17.1

